### DISCUSSION TITLE
APIs limitations for number of items returned

### DISCUSSION THREAD

**Discussion by DevLazio** [sreeise/graph-rs-sdk#360] [Ideas]
Currently, if you request an api which provide a list you may receive a response containing a next_link attribute. 
For example, 
`v1().users().list_user()` by default will only return 100 users, and a maximum of 999 can be ask with the `.top("999")` method.
So if you have more than a thousand users in your Azure AD you will need to use the next_link attribute to send another request for a new batch of users, append it to the previous list, rinse and repeat.
This means: 
- a lot a boilerplate code 
- new users of the crate may not be aware of the limitations before using their code on a production environnement

Should we do something about it ?
A first step could be to add some kind of warning in the documentation. Another step could be to provide a method to do that automatically. 

What's your take on this ?

**Comment by sreeise** [comment]
Yeah I think a good start here would be providing documentation on this.

After that I believe a method that can be called to request that the next link calls be made to the api and included in the response. We would probably have to do some type of channel if we don't want to block the thread getting the rest of those calls.

Example:

```rust
client
  .v1()
  .users()
  .list_users()
  .with_next_link_calls() // Tell the client you want to continue calling next link urls 
  .send()
```

I don't believe the crate should automatically call those, without explicit request, because doing so would go against the intended behavior of calling out to a single api endpoint. But I like the benefit this would bring to using the client.

**Comment by vladvasiliu** [comment]
> We would probably have to do some type of channel if we don't want to block the thread getting the rest of those calls

This could also be implemented with a stream. See, for example, the approach used in aws-sdk-rust: https://docs.rs/aws-sdk-ec2/latest/aws_sdk_ec2/client/fluent_builders/struct.DescribeInstances.html#method.into_paginator

You can then either make all the calls if the user explicitly asks that, or make them as the user iterates through the returned elements.

**Reply by sreeise** [reply]
Yeah for sure. Both options could be provided as well.

**Comment by sreeise** [comment]
Filed issue: https://github.com/sreeise/graph-rs/issues/362

**Comment by sreeise** [comment]
Feel free to contribute to the ticket as well if anyone would like too. I am currently pretty busy with work so its hard to find time to work on this right now, but I also welcome others to contribute to help move tickets like this one along faster.

**Reply by DevLazio** [reply]
Same here, but I hope to be able to work on that sometimes during the summer.

**Comment by DevLazio** [comment]
I'm not sure to understand why you would want to not block the thread waiting for the other pages. 
Are you talking about the async client, that should yield while waiting ?

I 100% agree about your `.with_next_link_calls()` approach but I think we could go farther and have a settings in the client similarily to the `timeout` setting.

As for the stream / paginator, that seems like a good idea.

**Reply by sreeise** [reply]
I am more referring to how the api calls will be made. So with a channel we can communicate on each api call.  There is probably not a good way not to block the main thread to be honest now that I am thinking about it.

Having a setting similar to timeout would be good as well.

**Comment by DevLazio** [comment]
So I did a bit of work, you can see it [here](https://github.com/DevLazio/graph-rs/tree/next_links).

I ran into some problems. In order to find and follow the next_link, we need to deserialize the response (in the case of the json() call). To do so, we need the <U> struct to NOT include the next_link but to be the "final" struct. For example, if you want to fetch `client.v1().users().list_user()` we need the struct to be User and not a struct containing a Vec<Users>.
This means : 
 - an API break
 - we can return directly a GraphResponse<Vec<U>>
 - the user doesn't have easy access to the other fields returned with the data (next_link, context, others ..?) and I don't know if that's something we should care about.
 - for `text()` and `bytes()` it's even another mess

Right now I don't know how to proceed anymore. On one hand I like the opportunity to get the whole list with 1 call and not having to declare 2 structs per list. On the other hand, I don't like API breaks and I'm not sure if it's the right direction to follow.
Maybe introduce this feature as an alternative to `json()`, like `json_with_next_links()` ?

What do you think ?

**Reply by vladvasiliu** [reply]
I've seen the same issue when taking a quick shot at implementing a stream.

I agree that breaking the API isn't great, but I always thought that the caller having to pass a type covering the full (instead of final) struct was a bit odd. The user's trying to get a list of Applications or whatever, not a list plus some other stuff.

> the user doesn't have easy access to the other fields returned with the data (next_link, context, others ..?) and I don't know if that's something we should care about.

I think that could be useful, at least for debugging purposes, but it doesn't seem incompatible.

I'm not overly familiar with the whole Graph API, so I may be wrong, but it seems to me that the outer response is basically always the same. So, the call could return a `GraphResponseBody<T>`, containing all the fields, among which `next_link`, etc., as well as `response_message: Vec<T>` which holds the actual expected objects.

[This link](https://docs.microsoft.com/en-us/graph/use-the-api#call-a-rest-api-method) seems to confirm that the "actual response" is always a (possibly missing) field, and that the outer response has the same generic fields.

**Reply by sreeise** [reply]
Still looking at the changes you posted so I will get back to you on that. But I do believe that the next link field is always in the top level object of the response body. There is a trait that was added a while back that provides a way to get common URL fields, such as next link. And these are common to all requests https://github.com/sreeise/graph-rs/blob/master/graph-http/src/traits/link.rs

https://github.com/sreeise/graph-rs/blob/fc9523c00dd319a01f8b4c7f43d4cc73c8b2f525/graph-http/src/traits/link.rs#L1-L30

The delta requests use this trait to get the next link and continue calling them. 
https://github.com/sreeise/graph-rs/blob/fc9523c00dd319a01f8b4c7f43d4cc73c8b2f525/graph-http/src/dispatch.rs#L273

I will try to respond to the changes you made here in the next couple of days but hopefully the above helps in some way.

**Reply by sreeise** [reply]
@DevLazio After looking at this some more and at your changes I think what might work best here is just to do something similar to the Delta requests I mentioned above. Basically how this will work is through a trait where you can require that any generic T for the response body implement a trait that has a method for the next link. Since the current ODataLink trait requires multiple url methods to implement you could probably create another trait that is for the callers to implement:

```rust
pub trait ODataNextLink<RHS = Self> {
    fn next_link(&self) -> Option<String>;
}
```

I have not tested that this works but as an example I believe it will be close to the delta requests:

```rust
    pub fn json_with_next_links<U: 'static + Send + ODataNextLink + Clone>(self) -> Receiver<Option<GraphResult<GraphResponse<U>>>>
        where
                for<'de> U: serde::Deserialize<'de>,
    {
        let (sender, receiver) = channel();
        if self.error.is_some() {
            sender
                .send(Some(Err(self.error.unwrap_or_default())))
                .unwrap();
            return receiver;
        }

        let initial_res: GraphResult<reqwest::blocking::Response> =
            self.request.send().map_err(GraphFailure::from);
        let response: GraphResult<GraphResponse<T>> = std::convert::TryFrom::try_from(initial_res);
        if let Err(err) = response {
            sender.send(Some(Err(e))).unwrap();
            return receiver;
        }

        let token = self.token;
        let response = response.unwrap();
        let mut next_link = response.body().next_link();
        sender.send(Some(Ok(response))).unwrap();

        thread::spawn(move || {
            let client = reqwest::blocking::Client::new();
            while let Some(next) = next_link {
                let res = client
                    .get(next.as_str())
                    .header(CONTENT_TYPE, "application/json")
                    .bearer_auth(token.as_str())
                    .send()
                    .map_err(GraphFailure::from);

                if let Err(err) = res {
                    next_link = None;
                    sender.send(Some(Err(err))).unwrap();
                } else {
                    match res.unwrap().with_graph_error() {
                        Ok(response) => {
                            let url = GraphUrl::from(response.url());
                            let headers = response.headers().clone();
                            let status = response.status();
                            match response.json::<T>() {
                                Ok(value) => {
                                    next_link = value.next_link();
                                    sender
                                        .send(Some(Ok(GraphResponse::new(
                                            url, value, status, headers,
                                        ))))
                                        .unwrap();
                                }
                                Err(err) => {
                                    next_link = None;
                                    sender.send(Some(Err(err.into()))).unwrap();
                                    is_done = true;
                                }
                            }
                        }
                        Err(err) => {
                            next_link = None;
                            sender.send(Some(Err(GraphFailure::GraphError(err)))).unwrap();
                        }
                    }
                }
            }
        });

        receiver
    }
```

Note that the delta requests use an enum to tell the caller when to stop listening to the channel. I didnt include something like that here and it may just require sending a None for the last sender call.

You will need to add other imports to intoresponse.rs as well for the above to work:

```rust
use graph_error::{WithGraphError, WithGraphErrorAsync};
use reqwest::header::CONTENT_TYPE;
use std::sync::mpsc::{channel, Receiver};
use std::thread;
```
There are some imports from those crates already at the top of intoresponse.rs so you probably just need to add these remaining to those imports.

But since you will need to implement this in dispatch.rs anyway you can just implement it once in dispatch.rs and from intoresponse.rs call the dispatch.rs method. This is also how its done for the delta requests.

**Reply by sreeise** [reply]
@vladvasiliu Id be interested to know if you have a solution where you are implementing streams for async requests? I think I tried it once and it was just difficult because the Rust standard library does not have streams for async yet and so you have to use other libraries implementations and its not always the easiest to do.

To be honest I havnt looked at using streams for blocking requests either so id be interesting to learn more about how we can do that as well.

**Reply by vladvasiliu** [reply]
@sreeise to be honest, I didn't spend too much time on this, and I was going to reach for tokio-stream, since the async part of this crate depends on tokio anyway. I hadn't considered the blocking use at all.

The solution is not yet usable, I got more or less to the point @DevLazio got: having to change the API for the response content. I was actually going to post a question along those lines, but wasn't fast enough.

I hadn't thought of your trait approach, which may actually not be so bad since it could be provided by the lib itself, seeing how responses seem pretty generic. So, the whole streaming thing could be built on top of the current one-shot approach, but wrapping it and providing this trait. My idea was finding a way to avoid having every user of the lib reimplement the same wrapper around the response.

I'll see if I can set some time aside next week to look into this.

**Reply by DevLazio** [reply]
Thanks a lots for your replies. I will look into it.
Having the trait does indeed remove the problem.

However I do agree with @vladvasiliu that forcing the caller to declare and provide a Struct that contains the type he wants feels weird, even with the trait. Maybe we could look into that later. (With some kind of derive macro ?)

**Reply by sreeise** [reply]
You could technically implement the trait for all T as long as you require it be `DeserializeOwned` or `for<'de> U: serde::Deserialize<'de>`. That will still require the caller to implement those for any custom type T but they would have to do that anyway to call the json method.

Derive Macros are def a solution that are easy to use in the long haul. Less easy to write but I have done them before. If you feel comfortable doing it then feel free to go for it and ill help in any way I can.

**Reply by DevLazio** [reply]
I implemented your solution @sreeise with a channel and it works : https://github.com/DevLazio/graph-rs/commit/d22f4563cc88fd6df11f79a10d0b551809dc84b9

But as a user I still think it's awkward to use because there is still the need to concatenate the pages together if you want to use them at the same time and you have to manage the receiver end of the channel.

So I pursued my "One time call" version, using your trait and I starting to like it. I only had to add a `value()` method to the trait, because I'm messing with the vec.

Here is the sample code I'm using to test the library. The sync part use the channel, the async part doesn't.

```rust
use graph_rs_sdk::http::GraphResponse;
use graph_rs_sdk::client::Graph;
use graph_rs_sdk::error::GraphResult;
use graph_rs_sdk::http::traits::{ODataNextLink, ODataNextLinkBlocking};

use serde::Deserialize;
use serde::Serialize;
use tokio::task::spawn_blocking;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct User {
    pub(crate) id: Option<String>,
    #[serde(rename = "userPrincipalName")]
    user_principal_name: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Users {
    pub(crate) value: Vec<User>,
    #[serde(rename = "@odata.nextLink")]
    pub(crate) next_link: Option<String>,
}

impl ODataNextLink<User> for Users {
    fn next_link(&self) -> Option<String> {
        return self.next_link.clone();
    }

    fn value(&mut self) -> &mut Vec<User> {
        return &mut self.value;
    }
}

impl ODataNextLinkBlocking for Users {
    fn next_link(&self) -> Option<String> {
        return self.next_link.clone();
    }
}


static ACCESS_TOKEN: &str = "";

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let client = Graph::new_async(ACCESS_TOKEN);

    let users_resp: GraphResponse<Users> = client
        .v1()
        .users()
        .list_user()
        .select(&["id", "userPrincipalName"])
        .top("100")
        .json()
        .await.unwrap();

    let users = users_resp.into_body();
    println!("{}", users.value.len());


    let users_resp: GraphResponse<Users> = client
        .v1()
        .users()
        .list_user()
        .select(&["id", "userPrincipalName"])
        .top("100")
        .with_next_link_calls()
        .json()
        .await.unwrap();

    let users = users_resp.into_body();
    println!("{}", users.value.len());
    spawn_blocking(|| main_sync());
    Ok(())
}


fn main_sync() {
    let client = Graph::new(ACCESS_TOKEN);

    let users_resp: GraphResponse<Users> = client
        .v1()
        .users()
        .list_user()
        .select(&["id", "userPrincipalName"])
        .top("100")
        .json()
        .unwrap();

    let users = users_resp.into_body();
    println!("{}", users.value.len());

    let foo = client
        .v1()
        .users()
        .list_user()
        .select(&["id", "userPrincipalName"])
        .top("100")
        .with_next_link_calls()
        .json_with_next_links::<Users>();

    let mut bar = foo.iter();
    while let Some(a) = bar.next() {
        if let Some(b) = a {
            println!("{:?}", b.unwrap().into_body().value.len());
        }
    }
}
```

**Reply by vladvasiliu** [reply]
Here's my first implementation of the stream. It's only implemented for the async part of the library.

https://github.com/sreeise/graph-rs/compare/248f929...vladvasiliu:graph-rs:f0f635c

A similar approach could probably be easily implemented for the blocking part.

**Reply by sreeise** [reply]
@DevLazio @vladvasiliu I like both your approaches to this. @DevLazio go ahead and open a PR when your ready with your changes. @vladvasiliu I feel like we could add your streaming approach to it. 

With streams we can make this more generic across multiple use cases. Lets start with @DevLazio approach and see if we can add onto it from there. That way we can let the caller decide if they want a stream, channel, or for the api to return the whole thing.

**Reply by sreeise** [reply]
```
 .with_next_link_calls()
.json_with_next_links::<Users>();
```

Maybe simplify this to just be 

```
json_with_next_links::<Users>();
```

Or with the option to do 3 different ways of getting the data we could have three methods.

```
.with_next_link_calls()
.stream()

.with_next_link_calls()
.channel()

.with_next_link_calls()
.json()
```

Up to yall though. Whatever feels best from a user perspective.

**Reply by vladvasiliu** [reply]
Not to bang my own drum, but I get the impression that building on the streams would be the most straightforward.

If the user wants them, great, they ask for them. If they want a channel, you can read from the stream and send the responses over. If they want the whole thing, you collect the stream in a Vec or something and return that.

On a more technical note, I see @DevLazio implemented following the `next_link` logic in the `IntoResponse`, whereas I implemented mine in the HTTP client - though that did give me the missing link for exposing this to the user.

On the one hand, I think that doing this is the client's business, but then again having to surface this through multiple layers feels like it might not be the best place. What's your take on this?

**Reply by sreeise** [reply]
1. I'm fine with leaving out a channel method in favor of streaming. 
2. You can have the logic in the http client and call it from within a method in `IntoResponse`. Thats pretty common. Butt 
3. I would prefer that we have some type of implementation that covers a multitude of use cases and not just http requests for streams. Or not just next links http requests at least. We do similar types of channel message passing on 
upload sessions.

**Reply by sreeise** [reply]
> Not to bang my own drum, but I get the impression that building on the streams would be the most straightforward.

Its not always that straight forward. Thats why I wanted to see and still do want to see DevLazio's PR first. 

I appreciate the unusual honesty I guess. Little bit of ego... Not sure if you meant for it to come off that way though 🤣.

**Reply by vladvasiliu** [reply]
It wasn't exactly an ego thing, more that "of course I like my solution better".

I agree that the implementation should be as general as possible.

> I would prefer that we have some type of implementation that covers a multitude of use cases and not just http requests for streams.

I'm not overly familiar with this lib. What requests besides HTTP are there?

> We do similar types of channel message passing on upload sessions.

OK, I'll have a look, not really sure what those are (I don't use MS Graph all that much).

**Reply by DevLazio** [reply]
I do think that streams are usefull and I might use them myself, especially if it's as simple as a method call for the user.
But sometimes I need the whole collection in one go at don't want to deal with the rest.
Letting the user choose the tool he wants to use is nice.

As far as my PR, I still need to do the blocking part. I'm also working on a proc macro to help the user craft his structs.
Once this is done I think we can do
```rust
.with_next_link_calls()
.json()
```
even if this is a (small) breaking change because it will be an improvement even for user who don't want to follow the next_links.

**Reply by vladvasiliu** [reply]
> But sometimes I need the whole collection in one go at don't want to deal with the rest.

I agree. Streams allow doing this easily, you can just `collect()` them like any Iterator. I think it's fair to say Streams are async Iterators.

https://docs.rs/futures-util/latest/futures_util/stream/trait.StreamExt.html#method.collect

**Reply by sreeise** [reply]
> > I would prefer that we have some type of implementation that covers a multitude of use cases and not just http requests for streams.
> 
> I'm not overly familiar with this lib. What requests besides HTTP are there?
> 

Not really non-HTTP requests. Just different ways of doing http requests. Downloading files is one where it becomes necessary to have some kind of delay and/or message passing to download large files or just multiple files.

**Reply by sreeise** [reply]
Having a stream method for different types would be nice to have that is for sure

**Reply by vladvasiliu** [reply]
> Having a stream method for different types would be nice to have that is for sure

By "different types" do you mean "Graph types", as in User / Application / etc? I think this could be easy to implement.

> Downloading files is one where it becomes necessary to have some kind of delay and/or message passing to download large files or just multiple files.

I think that could be useful, but I wonder whether it should be done at the same time with the `next_link` solution.

**Reply by sreeise** [reply]
By types I mean different types of http calls that may do different things. But this could also be said for returning different types as well such as in returning in bytes or returning in serde_json:;Value or a user defined type.

> I think that could be useful, but I wonder whether it should be done at the same time with the next_link solution.

Doesn't have to be. If we do it right then we can do something like a trait implementation that can be used for different api calls (and different types that can be used in those api calls).

**Reply by sreeise** [reply]
> Here's my first implementation of the stream. It's only implemented for the async part of the library.
> 
> [248f929...vladvasiliu:graph-rs:f0f635c](https://github.com/sreeise/graph-rs/compare/248f929...vladvasiliu:graph-rs:f0f635c)
> 
> A similar approach could probably be easily implemented for the blocking part.

@vladvasiliu Hey were you ever able to make some changes enough to open a PR for this? I def like your streaming approach. If you want to open a PR feel free, and if not I can just use the changes you posted before as a reference to implement something.

**Reply by vladvasiliu** [reply]
Hi @sreeise, I didn't have the time to look after this in a while. Would you like me to open a PR with only the async part, or would you rather wait to also have a blocking part at the same time?

If the former, I can probably whip something up.

**Reply by sreeise** [reply]
The async specific part is fine with me for sure. But after I wrote to ask you about it I just started looking at your implementation and I ended up just writing something that is based off how you used try_stream and stream. And I kind of reworked how it was done by moving the calls to stream and try_stream to intoresponse.rs and it seems to be working rather well. 

Sorry about that. I appreciate you looking at that originally. To be honest I am still a little confused on whats happening, or whats its doing under the hood, when calling try_stream inside the stream macro like you did so without your help it probably would have taken me a bit longer to get a working solution.

**Reply by sreeise** [reply]
This is what I added if your interested https://github.com/sreeise/graph-rs/pull/384

**Reply by vladvasiliu** [reply]
That's fine with me. I wasn't quite sure where the best place to implement that would be. To be honest, I'm a bit confused with the multiple layers between the user-facing interface and the actual HTTP client, since some functionality seems redundant at first look.

Basically, the idea of `try_stream!` is to have a stream that returns a `Result` so that it offers the same approach to interacting with the underlying HTTP calls. If I'm not mistaken, this isn't included in the proper tokio crate because the required functionality landed too late for the release of v1.

**Reply by sreeise** [reply]
Yeah I see what your saying. Its basically done the way it is in order to provide both the blocking and async client implementations. So there is some abstraction that is needed in order for both http clients to be bale to use the same api methods that are generated by macros. Otherwise it would require writing two separate implementations. Basically we cant call async methods from a blocking client because it doesnt have a runtime. So the api methods that are shared cannot be async at all.

**Reply by sreeise** [reply]
I will say though I have been thinking about just returning reqwest::Response for async requests and reqwest::blocking::Response for the blocking. For 90% of the requests this will work and wouldnt require so much work on converting the response and the body if there is one. There are a few cases, namely downloading files and upload sessions, where we would still need to have a solution similar to whats it is now.
